SMLHelper/Options/ConfigModOptions.cs

```diff
@@ -0,0 +1,952 @@
+                var modOptionEventFound = false;
+
+                for (var i = 0; i < parameterTypes.Length; i++)
+                {
+                    if (!senderFound && parameterTypes[i] == typeof(object))
+                    {
+                        senderFound = true;
+                        parameters[i] = sender;
+                    }
+                    else if (!eventArgsFound && parameterTypes[i] == typeof(TSource))
+                    {
+                        eventArgsFound = true;
+                        parameters[i] = e;
+                    }
+                    else if (!modOptionEventFound && parameterTypes[i] == typeof(IModOptionEventArgs))
+                    {
+                        modOptionEventFound = true;
+                        parameters[i] = e;
+                    }
+
+                    if (senderFound && eventArgsFound && modOptionEventFound)
+                        break;
+                }
+
+                memberInfoMetadata.InvokeMethod(Config, parameters);
+            }
+        }
+
+        /// <summary>
+        /// Generates tooltips for each <see cref="ModOption"/> with a specified <see cref="TooltipAttribute"/>, before
+        /// invoking any relevant method(s) specified with <see cref="OnGameObjectCreatedAttribute"/>(s) and passes
+        /// parameters when a <see cref="GameObject"/> is created in the options menu.
+        /// </summary>
+        /// <param name="sender"></param>
+        /// <param name="e"></param>
+        private void ConfigModOptions_GameObjectCreated(object sender, GameObjectCreatedEventArgs e)
+        {
+            if (modOptionsMetadata.TryGetValue(e.Id, out var modOptionMetadata))
+            {
+                // Create a tooltip if there is a TooltipAttribute specified
+                if (modOptionMetadata.ModOptionAttribute.Tooltip is string tooltip)
+                {
+                    e.GameObject.GetComponentInChildren<Text>().gameObject.AddComponent<ModOptionTooltip>().Tooltip = tooltip;
+                }
+
+                if (modOptionMetadata.OnGameObjectCreatedMetadata == null)
+                    return; // Skip attempting to invoke events if there are no OnGameObjectCreatedAttributes set for the member.
+
+                foreach (var onGameObjectCreatedMetadata in modOptionMetadata.OnGameObjectCreatedMetadata)
+                {
+                    InvokeEvent(onGameObjectCreatedMetadata, sender, e);
+                }
+            }
+        }
+        #endregion
+
+        #region Build ModOptions Menu
+        /// <summary>
+        /// Adds options to the menu based on the <see cref="modOptionsMetadata"/> dictionary, generated by the constructor.
+        /// </summary>
+        public override void BuildModOptions()
+        {
+            foreach (var entry in modOptionsMetadata
+                .OrderBy(x => x.Value.ModOptionAttribute.Order)
+                .ThenBy(x => x.Value.MemberInfoMetadata.Name))
```

Comment:
There are some concessions I am willing to make in the name of simplifying some of this code.  

Consider allowing yourself to make these assumptions (_and then document them in the guide for this feature_):  
1. All relevant members of the class will be either Fields _or_ properties, but never a mix of both
2. Entries show up in the options menu in the same order as they appear in the class
   - There would be no way for the modders to explicitly specify an order
   - The class must have its fields/properties set up in the desired order

Take these as suggestions.  
I understand wanting to provide modders with the most flexible tool possible, but at some point, constraining their options is actually better for everyone.

Reply:
You're forgetting methods (for buttons). There's no reason to make it so that they will always be at the bottom (or top).

When we get the update to .NET Standard, we'll be able to use `[CallerLineNumber]` to have the order work in order of declaration regardless of whether it's fields, properties, etc, quite easily.

Once we have that switch, we can get rid of the `static int` increment. As for setting the order yourself, it won't be too much effort for me to make it so that there is always a gap of `1` or `5` or `10` between declared members (rather than however many lines there are), so it becomes easy to insert things before auto-ordered ones.

However, that all said, I do agree that order of declaration is simplest implementation and I find it hard to justify why you would need to manually set the order in any other fashion, so I don't really mind the loss of the feature.